### ISSUE TITLE
NCEP Post and way of handling namelist file

### ISSUE THREAD

**Issue by uturuncoglu** [ufs-community/ufs-mrweather-app#65]
The current version of the NCEP Post namelist file (itag) is a combination of parameters that are read from standard input (stdin) and from namelist group called as nampgb. The problem is that the order of stdin ones are changing based on the selected options in the namelist file and this makes hard to manage the namelist file under CIME. As you already know, all namelist files (chgres, model and post) are generated by CIME on-the-fly and i think that the way of handling namelist file in NCEP post side is not the manageable for long term. So, i am proposing to define namelist options read from stdin as a different namelist group like nampgb by using fortran standards.

**Comment by ligiabernardet** [comment]
@fossell  Any concerns?

**Comment by junwang-noaa** [comment]
@WenMeng-NOAA, please see comments on itag from uturuncoglu

**Comment by WenMeng-NOAA** [comment]
DTC UPP team has been mainly working on UPP for ufs public release. It is okay for me if @fossell  has no concerns and the change will be added in UPP branch for ufs public release only.

**Comment by uturuncoglu** [comment]
@WenMeng-NOAA @junwang-noaa @fossell I have a workaround for it but it is not stable and if user make a change in some parameter it might fail. So, please let me know when you have those changes and i'll make the modification in CIME side.

**Comment by fossell** [comment]
@uturuncoglu Apologies for the delay, I was out of the office for a few days.  Can we discuss the changes you are proposing?  I'm not clear on exactly what you want changed.  Our team is not yet familiar with the CIME workflow or how post is run with this app, and our community release differs from operations in how the namelist file is generated, so it would be helpful to me if we could discuss or if you could give me more detail about what needs to change so we can be consistent.

**Comment by uturuncoglu** [comment]
@fossell No worries. It is not so urgent. Actually, the issue is not related with the CIME but the way of definition of namelist parameters that makes it hard to manage in a sustainable way.

I think that rather than using a namelist that includes both parameters read in stdin and fortran namelist together in the same file, it might be good to move all namelist options that are read in stdin to a new fortran namelist group. 

For example, look at fileNameFlat namelist option in sorc/ncep_post.fd/WRFPOST.f, it is read from unit=5 (stdin) as follows,

```
read(5,111,end=125) fileNameFlat
```

but it it could be defined a new namelist group like following (i just put fileNameFlat but there are others such as fileName, ioform, outform etc.

```
namelist/new_group/
fileNameFlat, ....
``` 

and then whole namelist group could be read as following like nampgb

```
read(5,new_group,iostat=iret,end=119)
```
This will enable us to read all namelist options without confusion in their order in the namelist file. BTW, i also could not understand why the way of definition of the same namelist option is changing based on the model name,

```
        if(MODELNAME == 'RAPR') then
          read(5,*,iostat=iret,end=119) kpo
        else
          read(5,nampgb,iostat=iret,end=119)
        endif
```

in this case, if model name is 'RAPR', it is read from stdin but in other case it is read from fortran namelist group. This kind of logicals makes hard to manage namelist file in a standardized way.

**Comment by fossell** [comment]
@uturuncoglu Thanks for clarifying, I see what you mean now.  In general I don't have concerns with this approach.  My reluctance right now is the timing of such a change.  The DTC is working hard to get some other final UPP changes into the ufs release branch in effort to ensure that the standalone community release the DTC will distribute in the coming month or so will be in sync with what is released as part of the UFS MRweather app.  The changes you propose would also require us to implement into our community release which is already undergoing final regression testing in preparation for release.   I'm concerned with the timeline of trying to put in your proposed changes which would require a good deal of extra testing and changes to documentation and procedures for our users.  We're also not adequately resourced right now to do this development and testing on a short timeline at the DTC, and it would be better suited for the post-release time frame.  So, while the current way of reading namelist parameters is admittedly convoluted, would it be ok to save this enhancement for the next release?  

The other thing to mention is that the UPP is currently undergoing a refactoring, in which some of these things may also be addressed, and for example some of the model name dependencies that you mention which are rampant throughout the code.  Given more time we could ensure that namelist modifications are inline with that refactoring effort as well.  

Thoughts?

**Comment by arunchawla-NOAA** [comment]
@uturuncoglu and @jedwards4b please let us know if this is satisfactory and this can be closed

**Comment by uturuncoglu** [comment]
@arunchawla-NOAA  @fossell Yes, i know the workload in EMC side and i have already implemented a workaround for this issue. So, it is not urgent. Just note that if user try to change the NCEP post namelist options that are read from stdio, the CIME workflow could fail or not get the changes.

**Comment by fossell** [comment]
@uturuncoglu Ok thanks.  And yes I understand that this will limit flexibility of users to modify namelist parameters.  But, I think for the first release it's ok if users do not have a fully customizable and flexible system, since those types of things won't be tested and officially supported.  Definitely something we want for the future though!

**Comment by fossell** [comment]
This request has been addressed by [UPP Issue 115](https://github.com/NOAA-EMC/UPP/issues/115) and merged into UPP develop branch with [UPP PR 402](https://github.com/NOAA-EMC/UPP/pull/402).  Closing this issue but please let me know if there are additional requests.
